contrib/babelfishpg_tsql/src/dbcmds.c

```diff
@@ -1231,3 +1259,343 @@ create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
+										 AccessShareLock);
+
+	user_name = (NameData *) palloc0(NAMEDATALEN);
+	snprintf(user_name->data, NAMEDATALEN, "%s", rolname);
+	ScanKeyInit(&key,
+				Anum_bbf_authid_user_ext_rolname,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				NameGetDatum(user_name));
+
+	scan = systable_beginscan(bbf_authid_user_ext_rel,
+				get_authid_user_ext_idx_oid(),
+				true, NULL, 1, &key);
+
+	tuple_user_ext = systable_getnext(scan);
+	if (HeapTupleIsValid(tuple_user_ext))
+		catalog_entry_exists = true;
+
+	systable_endscan(scan);
+	table_close(bbf_authid_user_ext_rel, AccessShareLock);
+	return catalog_entry_exists;
+}
+
+/* Grant permissions on all the existing objects to db_datareader/db_datawriter. */
+static void
+grant_perms_to_dbreader_dbwriter(const uint16 dbid,
+				const char *db_datareader,
+				const char *db_datawriter)
+{
+    Relation    namespace_rel;
+	TupleDesc   namespace_rel_descr;
+	ScanKeyData key;
+	HeapTuple   tuple;
+	TableScanDesc   tblscan;
+
+	namespace_rel = table_open(namespace_ext_oid, RowExclusiveLock);
+	namespace_rel_descr = RelationGetDescr(namespace_rel);
+
+	ScanKeyInit(&key,
+				Anum_namespace_ext_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+
+	tblscan = table_beginscan_catalog(namespace_rel, 1, &key);
+
+	tuple = heap_getnext(tblscan, ForwardScanDirection);
+
+	while (HeapTupleIsValid(tuple))
+	{
+		bool		isNull;
+		Datum		datum;
+		char		*schema_name;
+		char		*dbname = get_db_name(dbid);
+		StringInfoData	query;
+		List		*stmt_list;
+		Node		*stmts;
+		int			i = 0;
+		ListCell	*parsetree_item;
+		char		*schema_owner;
+		char		*dbo_user;
+		MigrationMode	baseline_mode = is_user_database_singledb(dbname) ? SINGLE_DB : MULTI_DB;
+
+		datum = heap_getattr(tuple, Anum_namespace_ext_namespace, namespace_rel_descr, &isNull);
+		schema_name = NameStr(*DatumGetName(datum));
+		schema_owner = GetUserNameFromId(get_owner_of_schema(schema_name), false);
+		dbo_user = get_dbo_role_name_by_mode(dbname, baseline_mode);
+
+		initStringInfo(&query);
+		appendStringInfo(&query, "GRANT dummy ON ALL TABLES IN SCHEMA dummy TO dummy; ");
+		appendStringInfo(&query, "GRANT dummy ON ALL TABLES IN SCHEMA dummy TO dummy; ");
+		appendStringInfo(&query, "GRANT dummy ON ALL TABLES IN SCHEMA dummy TO dummy; ");
+		appendStringInfo(&query, "GRANT dummy ON ALL TABLES IN SCHEMA dummy TO dummy; ");
+
+		/* Grant ALTER DEFAULT PRIVILEGES on schema owner and dbo user. */
+		appendStringInfo(&query, "ALTER DEFAULT PRIVILEGES FOR ROLE dummy, dummy IN SCHEMA dummy GRANT dummy ON TABLES TO dummy; ");
+		appendStringInfo(&query, "ALTER DEFAULT PRIVILEGES FOR ROLE dummy, dummy IN SCHEMA dummy GRANT dummy ON TABLES TO dummy; ");
+		appendStringInfo(&query, "ALTER DEFAULT PRIVILEGES FOR ROLE dummy, dummy IN SCHEMA dummy GRANT dummy ON TABLES TO dummy; ");
+		appendStringInfo(&query, "ALTER DEFAULT PRIVILEGES FOR ROLE dummy, dummy IN SCHEMA dummy GRANT dummy ON TABLES TO dummy; ");
```

Comment:
Is this to enable access for future objects?

Reply:
Yes.